src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java

```diff
@@ -184,103 +287,193 @@ public String getCredentialsId() {
     }
 
     @DataBoundSetter
-    public void setCredentialsId(String credentialsId) {
+    public void setCredentialsId(@CheckForNull String credentialsId) {
         this.credentialsId = Util.fixEmpty(credentialsId);
     }
 
-    @CheckForNull
-    public String getCheckoutCredentialsId() {
-        return checkoutCredentialsId;
+    @NonNull
+    public String getRepoOwner() {
+        return repoOwner;
     }
 
-    @DataBoundSetter
-    public void setCheckoutCredentialsId(String checkoutCredentialsId) {
-        this.checkoutCredentialsId = checkoutCredentialsId;
+    @NonNull
+    public String getRepository() {
+        return repository;
     }
 
-    public String getIncludes() {
-        return includes;
+    @NonNull
+    public String getServerUrl() {
+        return serverUrl;
     }
 
     @DataBoundSetter
-    public void setIncludes(@NonNull String includes) {
-        Pattern.compile(getPattern(includes));
-        this.includes = includes;
+    public void setServerUrl(@CheckForNull String serverUrl) {
+        this.serverUrl = BitbucketEndpointConfiguration.normalizeServerUrl(serverUrl);
     }
 
-    public String getExcludes() {
-        return excludes;
+    @NonNull
+    public List<SCMSourceTrait> getTraits() {
+        return Collections.unmodifiableList(traits);
     }
 
     @DataBoundSetter
-    public void setExcludes(@NonNull String excludes) {
-        Pattern.compile(getPattern(excludes));
-        this.excludes = excludes;
+    public void setTraits(@CheckForNull List<SCMSourceTrait> traits) {
+        this.traits = new ArrayList<>(Util.fixNull(traits));
     }
 
-    public String getRepoOwner() {
-        return repoOwner;
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("2.2.0")
+    @DataBoundSetter
```

Comment:
Prior to structs 1.8 all `@DataBoundSetter` fields with a non-default value would be reported from the pipeline snippetizer.

With structs 1.8, the `@Deprecated` ones will only be reported if they are required to be set.

If we remove `@DataBoundSetter` then any pipeline scripts out there that use the old fields will be broken, so to effectively change things we need to:

1. Upgrade structs to 1.8
2. Add `@Deprecated`
3. Retain `@DataBoundSetter`
4. Make the logic transparent to legacy code.

Additionally, I want consuming plugins to migrate to traits, so I add the `@Restricted` to force the eager migration to the new path (I could use the `@Deprecated` warning as a gentle nudge, but that will make removal of the old code paths harder when say going to version 3.x or 4.x)